### PR TITLE
[DRAFT] New internal development process.

### PR DESCRIPTION
**Pull request by pepol**
I would really like to start a discussion about the development process I've started to describe in this document.

 - How could the development process be improved?
 - Is it better to have a whole codebase in one repository or one repository per project?
 - Is there any way to rewrite the document for it to look better? What about reordering and reorganization?
 - Is there anything unclear about the ideas mentioned?
 - What are the biggest problems with this approach in your opinion?
 - Do you think it would be better to provide the freedom of choice of tools here or force everyone to use one specific site?
 - What kinds of tools could we provide here? Currently I have:
  - `check_style`, along with our version of `lint` (possibly)
  - `check_patch`, for patch formatting issues
  - maybe some git hooks for tags?

Please write any comments here. Also don't be afraid to push any changes to this repository.

Cc: @JakubNvk

### CODE REVIEW

**Comment by pepol** [comment]
Okay since the document is wirtten in a bad and somehow chaotic way, I will try to summarise the proposed changes here:

No teams structure should be enforced. Instead devs should be able to work on whichever projects they choose and switch any time without any bureaucracy...

Only organizational hierarchy enforced should be projects, each having one or more maintainers, that are responsible for the project's codebase. They can create a chain of lieutenants if they wish.

To allow a freedom of choice, it is proposed that instead of using feature branches, forks and opening pull requests just like here, the changes should be sent in textual form (patches, 1 commit = 1 pr = 1 patch) over email, where whole review will be carried out as well.

With that, `format-patch` `request-pull` and `am` git commands should be used. PR descriptions and commit messages are merged into one mechanism and several gitbhooks should be provided for bug tagging and marking them as fixed, etc.

Since the development channel is email, the review should be carried out on project's mailing lists, so there is an archive of all correspodence.

Afrer the reviewed code is accepted, it should be committed to mainline of project by maintainer, while keeping authorship information intact. There could be tools provided for different changelog viewing along with developer metrics.

The proposal could work with both, one repository as well as per-project repository set-up. However, per-project set-up is probably better. This is also open to discussion though.

**Comment by pepol** [comment]
[DRAFT v2]

**Comment by pepol** [comment]
The biggest open question is probably:
*Do we want to use one repository per whole (or almost whole) codebase or use smaller, separate repositories for each project?*

Now, there are pros and cons for both approaches (obviously), but IMO the biggest for each are:
<hr/>
**Single codebase:**
- easy to set-up infrastructure (one mailing list, one repository, one maintainer, no need for set-up beyond that)
- no need to subscribe/unsubscribe from mailing lists as developers move around the projects
- everything is available
- no need to make patches to multiple projects if making a big change and monitoring their progress/dependencies

<hr/>
**Per-project repositories:**
- less overhead (no need to download everything if we want to work on just one project)
- easier to find the place and person to send the patch to (no need for chain of trust and lieutenants)
- division of review and project management labour
- no central management and thus no single point of failure
- easy to add/remove project without affecting others or keeping them in history
- easy to change a maintainer

<hr/>
**Note:** Many of the cons of either approach could be possibly eliminated by some middle-ground, making a couple, not one, but few, big repositories that would hold a whole stack of topic projects. E.g. infrastructure repository for projects dealing with the whole infrastructure, etc.

Feel free to add any other pros for either solution. I would really appreciate if we could get a discussion about this going, since this is one of the defining moments of Imterra's whole internal culture.

**Comment by pepol** [comment]
[DRAFT v2.1] 
 - a small change about the non-requirement of `subsystem:` part of summary phrase, as well as a change of format of the summary phrase.
 - we no longer require the summary phrase to be a complete sentence with punctuation (a full stop at the end) and upper-case letters.
 - **note:** current changes are project-specific.

**Comment by JakubNvk** [comment]
*Do we want to use one repository per whole (or almost whole) codebase or use smaller, separate repositories for each project?*
I'm leaning more towards separate repos for each project.

*Is there any way to rewrite the document for it to look better? What about reordering and reorganization?*
I feel it's pretty well organized.

*What are the biggest problems with this approach in your opinion?*
Is it safe enough to mail code or will there be some internal mail?

*Do you think it would be better to provide the freedom of choice of tools here or force everyone to use one specific site?*
Depends. Will usage of one specific site help organazing and orientation in code? If so, then one specific site.

File: development-process/Guide [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+followed.
+
+The following paragraphs describe how the patch itself should be formatted.
+Note that, if you have your patches used in a git repository, proper patch
+formatting can be had with "git format-patch".  The tools cannot create the
+necessary text, though, so read the instructions below anyway.
+
+The canonical patch subject line is:
+
+  Subject: [PATCH 001/123] subsystem: summary phrase
```
Comment (JakubNvk) [inline development-process/Guide]: @pepol why `/123]`? Is this just an example?

Comment (pepol) [inline development-process/Guide]: Yes, example.

Comment (JakubNvk) [inline development-process/Guide]: Can it be in proper quotation marks?

**Comment by pepol** [comment]
> *What are the biggest problems with this approach in your opinion?*
> Is it safe enough to mail code or will there be some internal mail?

The mails would be, obviously, sent using internal company e-mail, with patches strictly forbidden to be sent to outside address. Since these would be also sent through SMTP then, the connection itself could be encrypted with TSL. Once the mail reaches our SMTP server, it won't be resent elsewhere, so it shouldn't travel in plain-text at all. However an encrypted VPN can be provided in order to be extra sure about this (same way the website would be accessed anyway if we go with that instead).

> *Do you think it would be better to provide the freedom of choice of tools here or force everyone to use one specific site?*
> Depends. Will usage of one specific site help organazing and orientation in code? If so, then one specific site.

I am not sure about this. However, I can imagine providing a website that would allow an easy viewing of code and commits that are already in some repository (git already provides this kind of website already, although it's just a basic viewing - `gitweb`), maybe even search in the code and other advanced features and integrate it into draken. However, I don't see any huge advantage of using website like this one (github) for all code review, since I find mail communication much easier and self-explanatory than all the buttons here.

My biggest selling point is probably the fact that basically everyone knows how to use e-mail, how to reply to specific point, etc., so why not use that for our advantage? Also, `git` itself is pretty much built to leverage mail-driven development (with `format-patch`, `am`, `send-email`, `imap-send` and others), so why not?

File: development-process/Guide (hunk 1 above)
Comment (pepol) [inline development-process/Guide]: I've reformulated the sentence, please take a look.

**Comment by JakubNvk** [comment]
>My biggest selling point is probably the fact that basically everyone knows how to use e-mail, how to reply to specific point, etc., so why not use that for our advantage?
True, allright.

@pepol LGTM then!

**Comment by pepol** [comment]
@JakubNvk I'm glad you like it but... most important thing that I want from you about this is probably: Can you imagine your set-up and how it would work with this? Because I don't want to take myself into account, since I already use mutt for one of my e-mail accounts, meaning that I'm pretty much set for using this, but others won't be as ready.

One of the most important things, for example, is that webmail clients (such as GMail) don't work with this, since pretty much all of them change whitespace in the mail's body (and also send multipart/* messages by default). Stand-alone clients (like Thunderbird) do this as well, usually, but can be set-up to work like you would expect them. So I guess this sort of "trouble" is worth the thought and discussion (more about specific clients can be found in [Linux developer's documentation](https://github.com/torvalds/linux/blob/master/Documentation/email-clients.txt)).

However if this doesn't seem like a problem to you, feel free to merge it, or give instructions on how you expect this to be squashed/rebased.

**Pull request merged**